MedallionShell.Tests/Streams/MergedLinesEnumerableTestBase.cs

```diff
@@ -98,40 +98,50 @@ void TestOneThrows(bool reverse)
         }
 
         [Test]
-        public void FuzzTest()
+        public async Task FuzzTest()
         {
             var pipe1 = new Pipe();
             var pipe2 = new Pipe();
 
-            var enumerable = new MergedLinesEnumerable(new StreamReader(pipe1.OutputStream), new StreamReader(pipe2.OutputStream));
+            var enumerable = this.Create(new StreamReader(pipe1.OutputStream), new StreamReader(pipe2.OutputStream));
 
             var strings1 = Enumerable.Range(0, 2000).Select(_ => Guid.NewGuid().ToString()).ToArray();
             var strings2 = Enumerable.Range(0, 2300).Select(_ => Guid.NewGuid().ToString()).ToArray();
 
-            void WriteStrings(IReadOnlyList<string> strings, Pipe pipe)
+            static void WriteStrings(IReadOnlyList<string> strings, Pipe pipe)
             {
                 var spinWait = default(SpinWait);
                 var random = new Random(Guid.NewGuid().GetHashCode());
-                using (var writer = new StreamWriter(pipe.InputStream))
+                using var writer = new StreamWriter(pipe.InputStream);
+                foreach (var line in strings)
                 {
-                    foreach (var line in strings)
+                    if (random.Next(5) == 1)
                     {
-                        if (random.Next(4) == 1)
-                        {
-                            spinWait.SpinOnce();
-                        }
-
-                        writer.WriteLine(line);
+                        spinWait.SpinOnce();
                     }
+
+                    writer.WriteLine(line);
                 }
             }
 
             var task1 = Task.Run(() => WriteStrings(strings1, pipe1));
             var task2 = Task.Run(() => WriteStrings(strings2, pipe2));
-            var consumeTask = Task.Run(() => enumerable.ToList());
-            Task.WaitAll(task1, task2, consumeTask);
+            Task.WaitAll(task1, task2);
 
```

Comment:
I think I had to swap the order for tests to pass. Is this a red flag?

Reply:
Yeah let's revert this change and make sure it still passes. Also, does this pass or fail on main? You didn't make any changes to Pipe I think so it may be an issue with the release branch.

Reply:
@madelson I looked more into this and gathered some numbers, but I'm lost as to how I should debug this.

Note: 
- Workarounds used in the test project to run tests on `master`:
  - updated  `PackageReference`s to `nunit` 3.13.3, `NUnit3TestAdapter` 3.17.0, `Microsoft.NET.Test.Sdk` 15.9.0
  - added `NU1902,NU1903` to `NoWarn` and set `CheckEolTargetFramework` to `false`

| chance \ Branch| `master`<br>(net46 / netcoreapp2.2) | `release-1.7`<br>(net462 / net6.0) | `AsyncEnumerable` with `Task.WaitAll(task1, task2, consumeTask)`<br>(net462 / net6.0) |
|--------|:--------:|:--------:|:--------:|
| 25% (current) |  516ms / 8s<br><img width="284" alt="image" src="https://github.com/madelson/MedallionShell/assets/7676354/50a62b4b-4db4-44cf-a20d-635720020e04"> | 511ms / 8.4s<br><img width="291" alt="image" src="https://github.com/madelson/MedallionShell/assets/7676354/7f367e43-eb42-4980-ba64-5d4f80784c2c"> |  522ms / timeout in the async case, 9.2s in the sync case<br><img width="289" alt="image" src="https://github.com/madelson/MedallionShell/assets/7676354/b6900371-f18e-4746-a4a4-7154185947b4">|
| 20% (new)      |  408ms / 6.4s<br><img width="285" alt="image" src="https://github.com/madelson/MedallionShell/assets/7676354/b5780998-e7de-47e8-99e0-0ee08a4bfd2e"> |  374ms / 6.2s<br> <img width="285" alt="image" src="https://github.com/madelson/MedallionShell/assets/7676354/241bd050-a84f-46f1-a6fb-b71de2b9a827">|  282ms / timeout in the async case, 5.9s in the sync case<br><img width="290" alt="image" src="https://github.com/madelson/MedallionShell/assets/7676354/b735cf3a-f1e8-4731-9dbf-ffb93af6e4b6">|

However, I noticed that a small change makes a difference.
1. If I start consume before waiting writes, I get a timeout:
```cs
            var consumeTask = Task.Run(enumerable.ToListAsync);
            Task.WaitAll(task1, task2);
```
2. **If I wait for writes before I start consuming, the performance is similar to the first two columns:**
```cs
            Task.WaitAll(task1, task2);
            var consumeTask = Task.Run(enumerable.ToListAsync);
```
- _This proves that the problem lies in `consumeTask`, not `SpinWait`._
3. If I do the same as but use much shorter `strings1` and `strings2`, the test completes within 20ms.
```cs
            // originally 2000
            var strings1 = Enumerable.Range(0, 20).Select(_ => Guid.NewGuid().ToString()).ToArray();
            // originally 2300
            var strings2 = Enumerable.Range(0, 23).Select(_ => Guid.NewGuid().ToString()).ToArray();
...
            // same as master or release-1.7
            var consumeTask = Task.Run(enumerable.ToListAsync);
            Task.WaitAll(task1, task2, consumeTask);

            CollectionAssert.AreEquivalent(strings1.Concat(strings2).ToList(), consumeTask.Result);

```
Therefore, I believe that my `IAsyncEnumerable` implementation is flawed in a way that somehow "explodes" for bigger inputs (the threshold fluctuates, but it's somewhere between 70 and low 100s).

Any idea how I should debug this? For one thing, I think replacing `Guid.NewGuid()` with a human-friendly value will help, but that's all I can think of.

Reply:
I also found that `consumeTask.Status` is `WaitingForActivation` even after a few seconds.

Reply:
Seems like ChatGPT came to the rescue again. (It wasn't helping a few hours ago.)

I lost its message, but it said something along the lines of "`StreamWriter` wasn't being disposed properly, causing the `Pipe`."

Now the test passes, but I can't run `spinWait.SpinOnce();` as often. Do you think the frequency should also be a `protected virtual` value?

Reply:
> I lost its message, but it said something along the lines of "StreamWriter wasn't being disposed properly, causing the Pipe's InputStream to wait indefinitely."

Makes sense; we have do dispose the writer to end the stream. Can you point me to the relevant code change?

> Now the test passes, but I can't run spinWait.SpinOnce(); as often. Do you think the frequency should also be a protected virtual value?

I'm not sure I follow here. As often as what? Does it fail when it runs more often? In what way? How does the overall time for this test case compare before and after the changes (I would expect it to be the same). Who would be overriding the frequency if it were protected virtual?

Reply:
@madelson
1. `StreamWriter` changes
[`FuzzTest`](https://github.com/madelson/MedallionShell/pull/109/files#diff-68fdbc9634d30b7e1a0bb438ab37b458f0c478766fba188c44ece72f93e41cacR102-R142) is the one where I had to change this. Note the parameter changes of the local function `WriteStrings`.
2. `spinWait.SpinOnce()` changes
In `FuzzTest`, specifically https://github.com/madelson/MedallionShell/pull/109/files#diff-68fdbc9634d30b7e1a0bb438ab37b458f0c478766fba188c44ece72f93e41cacR102-R121, I updated the `if` condition from `random.Next(4) == 1` to `random.Next(110) == 1`, so I'm spinning left often (25% -> 0.91%). If I use a greater value for `random.Next` (i.e. spin more often), then the test never ends (at least for like 30 minutes, after which I stop the test) _only in the async case_. Therefore I was wondering if `MergedLinesEnumerableTestAsync` and `MergedLinesEnumerableTestSync` should use different frequencies by overriding a `protected virtual` variable.

Reply:
@madelson Let me know if the above makes sense!

Reply:
@madelson Bumping this thread!

Reply:
> the test never ends (at least for like 30 minutes, after which I stop the test) only in the async case

This makes me feel like there is a bug somewhere. It could be in the MergedLinesEnumerable changes, it could be in the test code, or it could be in the `Pipe` code.

What I would suggest is to (temporarily) add some logging statements to the code like this:
```
static class TempLogger
{
    private static readonly object Lock = new();

    public static void Log(string message)
    {
       lock (Lock)
       {
            File.AppendAllLines(@"c:\dev\log.txt", [$"[{DateTime.Now}] {message}"]);
       }
    }
}
```

My assumption is that at some point we should stop seeing log statements as the code will enter a hung state. We can then add additional logs to try to get closer and closer to the point where each thread stops.

From there, hopefully we can deduce why it is hanging.